torchvision/transforms/v2/functional/_utils.py

```diff
@@ -1,8 +1,167 @@
+import functools
+import inspect
+import warnings
 from typing import Any
 
 import torch
+from torchvision import datapoints
 from torchvision.datapoints._datapoint import Datapoint
 
 
 def is_simple_tensor(inpt: Any) -> bool:
     return isinstance(inpt, torch.Tensor) and not isinstance(inpt, Datapoint)
+
+
+_KERNEL_REGISTRY = {}
+
+
+def _kernel_wrapper_internal(dispatcher, kernel):
+    dispatcher_params = list(inspect.signature(dispatcher).parameters)[1:]
+    kernel_params = list(inspect.signature(kernel).parameters)[1:]
+
+    needs_args_kwargs_handling = kernel_params != dispatcher_params
+
+    kernel_params = set(kernel_params)
+    explicit_metadata = {
+        input_type: available_metadata & kernel_params
+        for input_type, available_metadata in [(datapoints.BoundingBoxes, {"format", "canvas_size"})]
+    }
+
+    @functools.wraps(kernel)
+    def wrapper(inpt, *args, **kwargs):
+        input_type = type(inpt)
+
+        if needs_args_kwargs_handling:
+            # Convert args to kwargs to simplify further processing
+            kwargs.update(dict(zip(dispatcher_params, args)))
+            args = ()
+
+            # drop parameters that are not relevant for the kernel, but have a default value
+            # in the dispatcher
+            for kwarg in kwargs.keys() - kernel_params:
+                del kwargs[kwarg]
+
+            # add parameters that are passed implicitly to the dispatcher as metadata,
+            # but have to be explicit for the kernel
+            for kwarg in explicit_metadata.get(input_type, set()):
+                kwargs[kwarg] = getattr(inpt, kwarg)
+
+        output = kernel(inpt.as_subclass(torch.Tensor), *args, **kwargs)
+
+        if isinstance(inpt, datapoints.BoundingBoxes) and isinstance(output, tuple):
+            output, canvas_size = output
+            metadata = dict(canvas_size=canvas_size)
+        else:
+            metadata = dict()
+
+        return input_type.wrap_like(inpt, output, **metadata)
+
+    return wrapper
+
+
+def _register_kernel_internal(dispatcher, datapoint_cls, *, wrap_kernel=True):
+    registry = _KERNEL_REGISTRY.setdefault(dispatcher, {})
+    if datapoint_cls in registry:
+        raise TypeError(
+            f"Dispatcher '{dispatcher.__name__}' already has a kernel registered for type '{datapoint_cls.__name__}'."
+        )
+
+    def decorator(kernel):
+        registry[datapoint_cls] = _kernel_wrapper_internal(dispatcher, kernel) if wrap_kernel else kernel
+        return kernel
+
+    return decorator
+
+
+def register_kernel(dispatcher, datapoint_cls):
+    return _register_kernel_internal(dispatcher, datapoint_cls, wrap_kernel=False)
+
+
+def _noop(inpt, *args, __future_warning__=None, **kwargs):
```

Comment:
Why is this called `__future_warning__`?

Reply:
I wanted to make it clear that this thing is special and it doesn't collide with `kwargs`. But we can go for `msg` or the like as well. Your choice.